### PR TITLE
[exporter] Flip on queue batcher

### PR DESCRIPTION
**Pull request by sfc-gh-sili**
#### Description

This PR solves https://github.com/open-telemetry/opentelemetry-collector/issues/10368.

Previously we use a pushing model between the queue and the batch, resulting the batch size to be constrained by the `sending_queue.num_consumers`, because the batch cannot accumulate more than `sending_queue.num_consumers` blocked goroutines provide.

This PR changes it to a pulling model. We read from the queue until threshold is met or timeout, then allocate a worker to asynchronously send out the request.

#### Link to tracking issue
Fixes https://github.com/open-telemetry/opentelemetry-collector/issues/10368
https://github.com/open-telemetry/opentelemetry-collector/issues/8122

#### Testing

This PR swaps out `batch_sender` directly and still passes all the existing tests.

### CODE REVIEW

File: exporter/internal/queue/batcher.go [hunk 1]
```diff
@@ -94,9 +94,3 @@ func (qb *BaseBatcher) flushAsync(batchToFlush batch) {
 		qb.workerPool <- true
 	}()
 }
-
-// Shutdown ensures that queue and all Batcher are stopped.
-func (qb *BaseBatcher) Shutdown(_ context.Context) error {
-	qb.stopWG.Wait()
-	return nil
-}
```
Comment (bogdandrutu) [inline exporter/internal/queue/batcher.go:102]: Why this change?

File: exporter/internal/queue/default_batcher.go [hunk 2]
```diff
@@ -124,25 +124,19 @@ func (qb *DefaultBatcher) startTimeBasedFlushingGoroutine() {
...
-					qb.currentBatchMu.Unlock()
-					continue
-				}
-				batchToFlush := *qb.currentBatch
-				qb.currentBatch = nil
-				qb.currentBatchMu.Unlock()
-
-				// flushAsync() blocks until successfully started a goroutine for flushing.
-				qb.flushAsync(batchToFlush)
-				qb.resetTimer()
```
Comment (bogdandrutu) [inline exporter/internal/queue/default_batcher.go:138]: Not sure I understand this, can we do this in a separate PR?

File: exporter/internal/queue/batcher.go (hunk 1 above)
Comment (sfc-gh-sili) [reply exporter/internal/queue/batcher.go:102]: See the other comment.

File: exporter/internal/queue/default_batcher.go (hunk 2 above)
Comment (sfc-gh-sili) [reply exporter/internal/queue/default_batcher.go:138]: Thanks! Here it is: https://github.com/open-telemetry/opentelemetry-collector/pull/11666
`batch_sender_test` helped me detect that the original implementation is missing a flush on shutdown.

**Comment by sfc-gh-sili** [comment]
@dmitryax  Hi Dimitrii, I wonder if you know what would be a better way to make sure existing tests pass with both feature gate on and off. Manually enabling then disabling in every single exporter test could work but I wonder if there's other option

File: exporter/exporterhelper/internal/base_exporter.go [hunk 3]
```diff
@@ -21,9 +21,17 @@ import (
 	"go.opentelemetry.io/collector/exporter/exporterbatcher"
 	"go.opentelemetry.io/collector/exporter/exporterqueue" // BaseExporter contains common fields between different exporter types.
 	"go.opentelemetry.io/collector/exporter/internal"
+	"go.opentelemetry.io/collector/featuregate"
 	"go.opentelemetry.io/collector/pipeline"
 )
 
+var usePullingBasedExporterQueueBatcher = featuregate.GlobalRegistry().MustRegister(
+	"telemetry.UsePullingBasedExporterQueueBatcher",
+	featuregate.StageBeta,
```
Comment (dmitryax) [inline exporter/exporterhelper/internal/base_exporter.go:30]: Why starting with Beta? That sounds too aggressive. Let's start with Alpha

File: .chloggen/11637-exporter-queue-batcher.yaml [hunk 4]
```diff
@@ -0,0 +1,28 @@
+# Use this changelog template to create an entry for release notes.
+
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: breaking
```
Comment (dmitryax) [inline .chloggen/11637-exporter-queue-batcher.yaml:4]:
```suggestion
change_type: enhancement
```

File: exporter/exporterhelper/internal/base_exporter.go [hunk 5]
```diff
@@ -27,7 +27,7 @@ import (
 
 var usePullingBasedExporterQueueBatcher = featuregate.GlobalRegistry().MustRegister(
 	"telemetry.UsePullingBasedExporterQueueBatcher",
```
Comment (dmitryax) [inline exporter/exporterhelper/internal/base_exporter.go:29]:
```suggestion
	"exporter.UsePullingBasedExporterQueueBatcher",
```

File: .chloggen/11637-exporter-queue-batcher.yaml [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+# The name of the component, or a single word describing the area of concern, (e.g. otlpreceiver)
+component: exporterqueue
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Change exporter queue batching to use a pulling model.
```
Comment (dmitryax) [inline .chloggen/11637-exporter-queue-batcher.yaml:10]:
```suggestion
note: Introduce a feature gate exporter.UsePullingBasedExporterQueueBatcher to use the new pulling model in exporter queue batching.
```

File: exporter/exporterhelper/internal/queue_sender.go [hunk 7]
```diff
@@ -71,30 +72,49 @@ type QueueSender struct {
...
 		numConsumers:   numConsumers,
 		traceAttribute: attribute.String(ExporterKey, set.ID.String()),
 		obsrep:         obsrep,
 		exporterID:     set.ID,
 	}
-	consumeFunc := func(ctx context.Context, req internal.Request) error {
-		err := qs.NextSender.Send(ctx, req)
-		if err != nil {
-			set.Logger.Error("Exporting failed. Dropping data."+exportFailureMessage,
-				zap.Error(err), zap.Int("dropped_items", req.ItemsCount()))
```
Comment (dmitryax) [inline exporter/exporterhelper/internal/queue_sender.go:93]: Doesn't the function definition itself stay the same so we don't need to split it?

File: exporter/exporterhelper/internal/base_exporter.go [hunk 8]
```diff
@@ -27,7 +27,7 @@ import (
...
-	featuregate.StageBeta,
+	featuregate.StageAlpha,
 	featuregate.WithRegisterFromVersion("v0.114.0"),
```
Comment (dmitryax) [inline exporter/exporterhelper/internal/base_exporter.go:31]:
```suggestion
	featuregate.WithRegisterFromVersion("v0.115.0"),
```

File: exporter/exporterhelper/internal/queue_sender.go (hunk 7 above)
Comment (sfc-gh-sili) [reply exporter/exporterhelper/internal/queue_sender.go:93]: Good point. Thanks!

File: exporter/exporterhelper/internal/batch_sender_test.go [hunk 9]
```diff
@@ -158,7 +158,6 @@ func TestBatchSender_BatchExportError(t *testing.T) {
 			assert.Eventually(t, func() bool {
 				return sink.requestsCount.Load() == tt.expectedRequests &&
 					sink.itemsCount.Load() == tt.expectedItems &&
-					be.BatchSender.(*BatchSender).activeRequests.Load() == 0 &&
```
Comment (dmitryax) [inline exporter/exporterhelper/internal/batch_sender_test.go:161]: Why is this line removed? We should wait until there are no in-flight requests

**Review by dmitryax — APPROVED** [review]
One question about tests. Otherwise LGTM

File: exporter/exporterhelper/internal/batch_sender_test.go (hunk 9 above)
Comment (sfc-gh-sili) [reply exporter/exporterhelper/internal/batch_sender_test.go:161]: This is because if the new feature is turned on, we use `queueBatcher` instead of `BatchSender` so `BatchSender` is a nil. In the previous PR, I set up the test to run with both the feature on and off so this line has to be removed.

Comment (dmitryax) [reply exporter/exporterhelper/internal/batch_sender_test.go:161]: Ok sounds good 👍
